### PR TITLE
fix binomial function in fci_string

### PR DESCRIPTION
**Pull request by xubwa**
Fix issue #2105

### CODE REVIEW

File: pyscf/lib/mcscf/fci_string.c [hunk 1]
```diff
@@ -130,12 +130,7 @@ static int binomial(int n, int m)
                 if (m*2 <= n) {
                         m = n - m;
                 }
```
Comment (matthew-hennefarth) [inline pyscf/lib/mcscf/fci_string.c:132]: If you are using the recursion relation below, shouldn't these lines also be removed?

**Comment by matthew-hennefarth** [comment]
Generally speaking, recursion should be avoided as it is typically slow. Though I am not sure if the binomial function will be used in cases much beyond the current cache size due to CPU limitations to perform FCI calculations with such large number of orbitals/electrons, especially given the memoization used here.

To implement this more efficiently people often use a multiplicative version instead of a recursion. Of course, then one has to be careful of not producing over/underflowed values, and it is even more tricky to make sure we are not getting rounding errors due to some intermediate division. I had written an [integer only version in rust](https://github.com/matthew-hennefarth/sci-rs/blob/5b1c1a6ffcb4219fef67b3c62ce47006af589036/src/special/combinatorics.rs#L111-L115) which does not suffer from floating point precision. In c, and using the same input args as given in the binomial signature it might look something like

```c
static int binomial(int n, int m) {
    if (m >= n) {
        return 1;
    } 
    if (n <= 20) {
        return _binomial_cache[n*(n-1)/2+m];
    } 
    int k = n + 1;
    int n_terms = min(m, n-m) + 1; // supposing a min function is defined somewhere...

    int result = 1;
    for (int i = 1; i < n_terms; i++) {
        result = (result * (k-i)) /i;
    }
    return result;
}
```

Not saying you should change the implementation to this version, but just something to keep in mind...

**Comment by xubwa** [comment]
I used a recursion here because in practice, a fci calculation with many electrons and many orbitals is impossible, it either exceeds the cached binomial array up to 21/22 orbitals by a few orbitals or the system only have very few electrons, yielding few recursion layers and can also make full use of the cahced array to get some savings. The code you provide seems able to guarantee the result to be integer at every step, so it also looks good to me. I have no strong preference on this as long as the problem can be fixed.

**Comment by xubwa** [comment]
I used a recursion here because in practice, a fci calculation with many electrons and many orbitals is impossible, it either exceeds the cached binomial array up to 21/22 orbitals by a few orbitals or the system only have very few electrons, yielding few recursion layers and can also make full use of the cahced array to get some savings. The code you provide seems able to gaurantee

**Comment by matthew-hennefarth** [comment]
> I used a recursion here because in practice, a fci calculation with many electrons and many orbitals is impossible, it either exceeds the cached binomial array up to 21/22 orbitals by a few orbitals or the system only have very few electrons, yielding few recursion layers and can also make full use of the cahced array to get some savings.

I agree, this implementation should be fine for PySCF.

**Review by sunqm — APPROVED** [review]

**Pull request merged**